statsmodels/base/_penalties.py

```diff
@@ -73,54 +86,457 @@ def grad(self, params):
+    see docstring of SCAD
+
+    Parameters
+    ----------
+    tau : float
+        slope and threshold for linear segment
+    c : float
+        factor for second threshold
+    c0 : float
+        threshold for quadratically smoothed segment
+    restriction : None or array
+        linear constraints for
+
+
+    Notes
+    -----
+    TODO: Use delegation instead of subclassing, so smoothing can be added to
+    all penalty classes.
+
+    """
+
+    def __init__(self, tau, c=3.7, c0=None, weights=None, restriction=None):
+        if weights is None:
+            self.weights = 1.
+        else:
+            self.weights = weights
+        self.alpha = 1.
+        self.tau = tau
+        self.c = c
+        self.c0 = c0 if c0 is not None else tau * 0.1
+        if self.c0 > tau:
+            raise ValueError('c0 cannot be larger than tau')
+
+        # get coefficients for quadratic approximation
+        c0 = self.c0
+        deriv_c0 = super(SCADSmoothed, self).grad(c0)
+        value_c0 = super(SCADSmoothed, self).func(c0)
+        self.aq1 = value_c0 - 0.5 * deriv_c0 * c0
+        self.aq2 = 0.5 * deriv_c0 / c0
+        self.restriction = restriction
+
+    def func(self, params):
+        # TODO: `and np.size(params) > 1` is hack for llnull, need better solution
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        # Note: we have the same problem with `restriction`
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).func(params[None, ...])
+        self.weights = weights
+
+        # shift down so func(0) == 0
+        value -= self.aq1
+        #change the segment corrsponding to quadratic approximation
+        p_abs = np.atleast_1d(np.abs(params))
+        mask = p_abs < self.c0
+        p_abs_masked = p_abs[mask]
+        value[mask] = self.aq2 * p_abs_masked**2
+
+        return (self.weights * value).sum(0)
+
+    def grad(self, params):
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).grad(params)
+        self.weights = weights
+
+        #change the segment corrsponding to quadratic approximation
+        p = np.atleast_1d(params)
+        mask = np.abs(p) < self.c0
+        value[mask] = 2 * self.aq2 * p[mask]
+
+        if self.restriction is not None and np.size(params) > 1:
+            return weights * value.dot(self.restriction)
+        else:
+            return weights * value
+
+    def deriv2(self, params):
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
```

Comment:
Not at all necessary here, but a pattern I've been liking for this kind of thing is to put these temporarily-patch-attribute blocks inside a try/finally block.

Reply:
I don't like `try...` blocks most of the time.
During development, i.e. as long as code is not settled, I prefer exceptions for any cases that are not handled yet.
In the final code I find `try` blocks usually not explicit enough. I like to see from the code what conditions need to be satisfied, i.e. why the try might not work, unless we catch something generic like convergence failures or missing internet access or failing imports.